### PR TITLE
smb: check file lengths against nbss records length

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5770

Describe changes:
- smb : fix evasions by abusing length field

I still need to create a pcap with this attack...

suricata-verify-pr: 1049

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 103936 | 81.05% |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 2845 | 0.00 | 

Pipeline 11314

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 103984 | 81.09% |
| | TREX_GENERIC_stats_chk | | |
| .tcp.segment_from_pool | 15728353 | 16631957 | 105.75% | 

Pipeline 11315

File: rust/src/smb/smb1.rs [hunk 1]
```diff
@@ -958,6 +958,14 @@ pub fn smb1_write_request_record<'b>(state: &mut SMBState, r: &SmbRecord<'b>, an
                 None => b"<unknown>".to_vec(),
             };
             let mut set_event_fileoverlap = false;
+            let max_rd_len = if rd.len as usize > rd.data.len() + more_nbss {
+                // Write record claims more bytes than can be contained in NBSS record...
+                // The server MUST fail the request with STATUS_INVALID_PARAMETER.
+                // We use tha data available...
```
Comment (victorjulien) [inline rust/src/smb/smb1.rs:964]: typo `tha`

File: rust/src/smb/smb1.rs [hunk 2]
```diff
@@ -1032,13 +1040,21 @@ pub fn smb1_read_response_record<'b>(state: &mut SMBState, r: &SmbRecord<'b>, an
             Ok((_, rd)) => {
                 SCLogDebug!("SMBv1: read response => {:?}", rd);
 
+                let max_rd_len = if rd.len as usize > more_nbss + rd.data.len() {
+                    // Write record claims more bytes than can be contained in NBSS record...
+                    // The server MUST fail the request with STATUS_INVALID_PARAMETER.
+                    // We use tha data available...
```
Comment (victorjulien) [inline rust/src/smb/smb1.rs:1046]: typo `tha`

File: rust/src/smb/smb1.rs [hunk 3]
```diff
@@ -1023,7 +1031,7 @@ pub fn smb1_write_request_record<'b>(state: &mut SMBState, r: &SmbRecord<'b>, an
     smb1_request_record_generic(state, r, events);
 }
 
-pub fn smb1_read_response_record<'b>(state: &mut SMBState, r: &SmbRecord<'b>, andx_offset: usize)
+pub fn smb1_read_response_record<'b>(state: &mut SMBState, r: &SmbRecord<'b>, andx_offset: usize, more_nbss: usize)
```
Comment (victorjulien) [inline rust/src/smb/smb1.rs:1034]: I find the `more_nbss` variable name not clear.

File: rust/src/smb/smb2.rs [hunk 4]
```diff
@@ -122,21 +122,29 @@ pub fn smb2_read_response_record<'b>(state: &mut SMBState, r: &Smb2Record<'b>)
 
     match parse_smb2_response_read(r.data) {
         Ok((_, rd)) => {
+            let max_rd_len = if rd.len as usize - rd.data.len() > more_nbss {
+                // Write record claims more bytes than can be contained in NBSS record...
+                // The server MUST fail the request with STATUS_INVALID_PARAMETER.
+                // We use tha data available...
```
Comment (victorjulien) [inline rust/src/smb/smb2.rs:128]: typo `tha`

File: rust/src/smb/smb2.rs [hunk 5]
```diff
@@ -279,10 +287,18 @@ pub fn smb2_write_request_record<'b>(state: &mut SMBState, r: &Smb2Record<'b>)
     }
     match parse_smb2_request_write(r.data) {
         Ok((_, wr)) => {
-            if (state.max_write_size != 0 && wr.wr_len > state.max_write_size) ||
-               (unsafe { SMB_CFG_MAX_WRITE_SIZE != 0 && SMB_CFG_MAX_WRITE_SIZE < wr.wr_len }) {
+            let max_wr_len = if wr.wr_len as usize - wr.data.len() > more_nbss {
+                // Write record claims more bytes than can be contained in NBSS record...
+                // The server MUST fail the request with STATUS_INVALID_PARAMETER.
+                // We use tha data available...
```
Comment (victorjulien) [inline rust/src/smb/smb2.rs:293]: typo `tha`

**Review by victorjulien — CHANGES_REQUESTED** [review]
git commit typo: leength

if we detect a bad record, we should set an event too, right?

File: rust/src/smb/smb1.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/smb/smb1.rs:1046]: fixed

File: rust/src/smb/smb1.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/smb/smb1.rs:964]: fixed

File: rust/src/smb/smb2.rs (hunk 4 above)
Comment (catenacyber) [reply rust/src/smb/smb2.rs:128]: fixed

File: rust/src/smb/smb2.rs (hunk 5 above)
Comment (catenacyber) [reply rust/src/smb/smb2.rs:293]: And fixed

By the way, lol : `git grep "tha "`

File: rust/src/smb/smb1.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/smb/smb1.rs:1034]: Proposing `nbss_remaining`
Do you have a better name ?

**Comment by catenacyber** [comment]
> git commit typo: leength

Fixed

> if we detect a bad record, we should set an event too, right?

Right, reusing `WriteRequestTooLarge` and `ReadResponseTooLarge` and updating their description which was kind of incomplete already...

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/8351

**Pull request closed without merging**